### PR TITLE
feat: Initial Plugin infrastructure for Application creation

### PR DESCRIPTION
**Pull request by e4r7hbug**
@sijis has a fairly stable Branch going for the new Plugin system. This should be a relatively safe switch over.

### CODE REVIEW

**Comment by sijis** [comment]
@jvasallo @sbasgall @darinpantley Do you have any concerns/issues with merging this in?

It should not affect any app creation process besides adding another parameter named 'provider'.

File: src/foremast/app/aws.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+"""AWS Spinnaker Application."""
+from pprint import pformat
+
+from foremast.app import base
+from foremast.utils import wait_for_task
+
+
+class SpinnakerApp(base.BaseApp):
+    """Create AWS Spinnaker Application."""
```
Comment (jvasallo) [inline src/foremast/app/aws.py]: Is this still true or can provider be modified to support various apps

**Comment by jvasallo** [comment]
I guess I am ok with the Pull Request, however, what is the end goal of this? Is this a way to extend foremast allowing foremast to be extended via various plugins and such (for example support for k8s, gcp)?

**Comment by savilesupgrade** [comment]
Rebased by popular demand. :)

**Comment by jvasallo** [comment]
Looks good to me. This is a major change so lets make this the dawn of the 4.x release.
@foremast/gogo @foremast/core @foremast/upgrade @foremast/redbox FYI

**Pull request merged**